File: 251001/Voytov/src/main/java/org/ex/distributed_computing/service/NoticeService.java
package org.ex.distributed_computing.service;

import java.util.List;
import lombok.RequiredArgsConstructor;
import org.ex.distributed_computing.dto.request.NoticeRequestDTO;
import org.ex.distributed_computing.dto.response.NoticeResponseDTO;
import org.ex.distributed_computing.exception.NotFoundException;
import org.ex.distributed_computing.mapper.NoticeMapper;
import org.ex.distributed_computing.model.News;
import org.ex.distributed_computing.model.Notice;
import org.ex.distributed_computing.repository.NewsRepository;
import org.ex.distributed_computing.repository.NoticeRepository;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class NoticeService {

  private final NoticeRepository noticeRepository;
  private final NewsRepository newsRepository;
  private final NoticeMapper noticeMapper;

  public List<NoticeResponseDTO> getAllNotices() {
    return noticeMapper.toDtoList(noticeRepository.findAll());
  }

  public NoticeResponseDTO getNoticeById(Long id) {
    Notice notice = noticeRepository.findById(id)
        .orElseThrow(() -> new NotFoundException("Notice not found"));
    return noticeMapper.toDto(notice);
  }

  public NoticeResponseDTO createNotice(NoticeRequestDTO requestDTO) {
    News news = newsRepository.findById(requestDTO.newsId())
        .orElseThrow(() -> new NotFoundException("News not found"));

    Notice notice = new Notice(news, requestDTO.content());
    noticeRepository.save(notice);
    return noticeMapper.toDto(notice);
  }

  public NoticeResponseDTO updateNotice(NoticeRequestDTO requestDTO) {
    Notice notice = noticeRepository.findById(requestDTO.id())
        .orElseThrow(() -> new NotFoundException("Notice not found"));

    News news = newsRepository.findById(requestDTO.newsId())
        .orElseThrow(() -> new NotFoundException("News not found"));

    notice.setNews(news);
    notice.setContent(requestDTO.content());
    noticeRepository.save(notice);
    return noticeMapper.toDto(notice);
  }

  public void deleteNotice(Long id) {
    Notice notice = noticeRepository.findById(id)
        .orElseThrow(() -> new NotFoundException("Notice not found"));
    noticeRepository.delete(notice);
  }
}

File: 251001/Voytov/discussion/src/main/kotlin/org/ex/discussion/service/NoticeService.kt
package org.ex.discussion.service

import org.ex.discussion.dto.request.NoticeRequestDTO
import org.ex.discussion.dto.response.NoticeResponseDTO
import org.ex.discussion.exception.NotFoundException
import org.ex.discussion.mapper.NoticeMapper
import org.ex.discussion.repository.NoticeRepository
import org.springframework.stereotype.Service

@Service
class NoticeService(
    private val noticeMapper: NoticeMapper,
    private val noticeRepository: NoticeRepository) {

    fun getAllNotices(): List<NoticeResponseDTO> = noticeRepository.findAll().map { noticeMapper.toDto(it) }

    fun getNoticeById(noticeId: Long): NoticeResponseDTO {
        return noticeRepository.findById(noticeId)
            .map { noticeMapper.toDto(it) }
            .orElseThrow {throw NotFoundException("Notice with id: $noticeId not found")}
    }

    fun createNotice(noticeRequest: NoticeRequestDTO): NoticeResponseDTO {
        val notice = noticeMapper.toEntity(noticeRequest)
        noticeRepository.save(notice)
        return noticeMapper.toDto(notice)
    }

    fun updateNotice(noticeRequest: NoticeRequestDTO): NoticeResponseDTO {
        return noticeRepository.findById(noticeRequest.id)
            .map {
                it.content = noticeRequest.content
                it.newsId = noticeRequest.newsId
                val updated = it.copy(content = noticeRequest.content, newsId = noticeRequest.newsId)
                return@map noticeMapper.toDto(noticeRepository.save(updated))
            }.orElseThrow {throw NotFoundException("Notice with id: ${noticeRequest.id}")}
    }

    fun deleteNotice(noticeId: Long): NoticeResponseDTO {
        return noticeRepository.findById(noticeId)
            .map {
                noticeRepository.delete(it.id, it.country)
                return@map it
            }.map { noticeMapper.toDto(it) }
            .orElseThrow{ throw NotFoundException("Notice with id: $noticeId not found") }
    }
}